components/sync15/src/state.rs

```diff
@@ -73,6 +69,94 @@ impl Default for PersistedGlobalState {
     }
 }
 
+#[derive(Debug, Default, Clone, PartialEq)]
+pub(crate) struct EngineChangesNeeded {
+    pub local_resets: HashSet<String>,
+    pub remote_wipes: HashSet<String>,
+}
+
+#[derive(Debug, Default, Clone, PartialEq)]
+struct RemoteEngineState {
+    info_collections: HashSet<String>,
+    declined: HashSet<String>,
+}
+
+#[derive(Debug, Default, Clone, PartialEq)]
+struct EngineStateInput {
+    local_declined: HashSet<String>,
+    remote: Option<RemoteEngineState>,
+    user_changes: HashMap<String, bool>,
+}
+
+#[derive(Debug, Default, Clone, PartialEq)]
+struct EngineStateOutput {
+    // The new declined.
+    declined: HashSet<String>,
+    // Which engines need resets or wipes.
+    changes_needed: EngineChangesNeeded,
+}
+
+fn compute_engine_states(input: EngineStateInput) -> EngineStateOutput {
+    use crate::util::*;
+    log::debug!("compute_engine_states: input {:?}", input);
+    let (must_enable, must_disable) = partition_by_value(&input.user_changes);
+    let have_remote = input.remote.is_some();
+    let RemoteEngineState {
+        info_collections,
+        declined: remote_declined,
+    } = input.remote.clone().unwrap_or_default();
+
+    let both_declined_and_remote = set_intersection(&info_collections, &remote_declined);
+    if !both_declined_and_remote.is_empty() {
+        // Should we wipe these too?
```

Comment:
I think so, yes, but we can save that for later, like you noted in the comment above `remote_wipes`. Mind filing an issue?

Reply:
Filed #1799